tests/chainer_tests/links_tests/connection_tests/test_peephole.py

```diff
@@ -0,0 +1,211 @@
+        h_shape = (1, self.out_size)
+        x_shape = (4, self.in_size)
+        gy_shape = (4, self.out_size)
+        self.c = numpy.zeros(c_shape).astype(numpy.float32)
+        self.h = numpy.zeros(h_shape).astype(numpy.float32)
+        self.x = numpy.random.uniform(-1, 1, x_shape).astype(numpy.float32)
+        self.gy = numpy.random.uniform(-1, 1, gy_shape).astype(numpy.float32)
+
+    def _forward(self, link, x):
+        return link(x)
+
+    def check_forward(self, c_data, h_data, x_data):
+        x = chainer.Variable(x_data)
+
+        h1 = self.link(x)
+        c1_expect, h1_expect = _peephole(self.link, c_data, h_data, x_data)
+        gradient_check.assert_allclose(h1.data, h1_expect)
+        gradient_check.assert_allclose(self.link.c.data, c1_expect)
+        gradient_check.assert_allclose(self.link.h.data, h1_expect)
+
+        h2 = self.link(x)
+        c2_expect, h2_expect = _peephole(self.link,
+                                         c1_expect, h1_expect, x_data)
+        gradient_check.assert_allclose(h2.data, h2_expect)
+        gradient_check.assert_allclose(self.link.c.data, c2_expect)
+        gradient_check.assert_allclose(self.link.h.data, h2_expect)
+
+    def test_forward_cpu(self):
+        self.check_forward(self.c, self.h, self.x)
+
+    @attr.gpu
+    def test_forward_gpu(self):
+        self.link.to_gpu()
+        self.check_forward(cuda.to_gpu(self.c),
+                           cuda.to_gpu(self.h),
+                           cuda.to_gpu(self.x))
+
+    def check_backward(self, c_data, h_data, x_data, y_grad):
+        x = chainer.Variable(x_data)
+        y = self._forward(self.link, x)
+        y.grad = y_grad
+        y.backward()
+
+        def f():
+            c, y = _peephole(self.link, c_data, h_data, x_data)
+            return y,
+        gx, = gradient_check.numerical_grad(f, (x.data,), (y.grad,))
+        gradient_check.assert_allclose(gx, x.grad, atol=1e-3)
+
+    def test_backward_cpu(self):
+        self.check_backward(self.c, self.h, self.x, self.gy)
+
+    @attr.gpu
+    def test_backward_gpu(self):
+        self.link.to_gpu()
+        self.check_backward(cuda.to_gpu(self.c),
+                            cuda.to_gpu(self.h),
+                            cuda.to_gpu(self.x),
+                            cuda.to_gpu(self.gy))
+
+
+@testing.parameterize(
+    *testing.product({
+        'link_array_module': ['to_cpu', 'to_gpu'],
+        'state_array_module': ['to_cpu', 'to_gpu']
+    }))
+class TestPeepholeState(unittest.TestCase):
+
+    def setUp(self):
+        in_size, out_size = 10, 8
+        self.link = links.Peephole(in_size, out_size)
+
+    def check_reset_state(self):
+        self.link.reset_state()
+        self.assertIsNone(self.link.c)
+        self.assertIsNone(self.link.h)
+
+    def test_reset_state_cpu(self):
+        self.check_reset_state()
+
+    @attr.gpu
+    def test_reset_state_gpu(self):
+        getattr(self.link, self.link_array_module)()
```

Comment:
I noticed that I did not have to use `self.link_array_module` because CPU test was done in ` test_reset_state_cpu`. So I did not remove `@attr.gpu` but change `getattr(self.link, self.link_array_module)() to self.link.to_gpu()`